server/src/main/java/org/candlepin/controller/ActualRefresher.java

```diff
@@ -0,0 +1,1297 @@
+
+        */
+
+        // Check if we have changes
+        if (node.changed()) {
+            if (!node.hasExistingEntity()) {
+                // Creation (5, 6)
+
+                switch (node.getType()) {
+                    case PRODUCT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Product) null,
+                            (Product) node.getImportedEntity()));
+
+                        break;
+
+                    case CONTENT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Content) null,
+                            (Content) node.getImportedEntity()));
+
+                        break;
+
+                    default:
+                        throw new IllegalStateException("Unexpected node type: " + node.getType());
+                }
+            }
+            else {
+                // Update (1, 2)
+
+                switch (node.getType()) {
+                    case PRODUCT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Product) node.getExistingEntity(), (ProductInfo) node.getImportedEntity()));
+
+                        break;
+
+                    case CONTENT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Content) node.getExistingEntity(), (ContentInfo) node.getImportedEntity()));
+
+                        break;
+
+                    default:
+                        throw new IllegalStateException("Unexpected node type: " + node.getType());
+                }
+            }
+        }
+        else if (childrenUpdated) {
+            node.markChanged(); // probably unnecessary, but might as well be consistent
+
+            // Resolve children references
+            switch (node.getType()) {
+                case PRODUCT:
+                    node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                        (Product) null,
+                        (Product) node.getImportedEntity()));
+
+                    break;
+
+                case CONTENT:
+                    node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                        (Content) null,
+                        (Content) node.getImportedEntity()));
+
+                    break;
+
+                default:
+                    throw new IllegalStateException("Unexpected node type: " + node.getType());
+            }
+        }
+
+        // Flag the node as visited so we don't risk rework or making changes multiple
+        // times, screwing up our entity refs
+        node.markVisited();
+    }
+
+
+    private Product createEntity(NodeMapper mapper, VersionMapper vMap,
+        Product existingEntity, ProductInfo importedEntity) {
+
+        Product updatedEntity = existingEntity != null ?
+            (Product) existingEntity.clone() :
+            new Product();
```

Comment:
In the case we're creating a new instance (instead of cloning), we also need to set the id field on it.

Reply:
Yeah, forgot that bit.